ext/DDProf/src/pprof_exporter.cc

```diff
@@ -0,0 +1,230 @@
+    std::vector<T *> vector;
+    uint64_t id;
+
+  public:
+    explicit pprof_set(datadog_arena **a) : arena{a}, id{0} {}
+    offset_t insert(T &elem) {
+        auto iterator = map.find(&elem);
+        if (iterator == map.end()) {
+            auto size = (offset_t) vector.size();
+            auto inserted = (T *) datadog_arena_alloc(arena, sizeof(T));
+            new(inserted) T(elem);
+            inserted->id = id++;
+
+            map[inserted] = size;
+            vector.emplace_back(inserted);
+            return size;
+        } else {
+            return iterator->second;
+        }
+    }
+
+    size_t size() noexcept { return vector.size(); }
+    T &operator[](size_t offset) noexcept { return *vector[offset]; }
+    T **data() noexcept { return vector.data(); }
+};
+
+uint64_t add_function(datadog_arena **arena, pprof_set<PProf__Function> &functions, uint64_t function_name) noexcept {
+    PProf__Function function = PPROF__FUNCTION__INIT;
+    function.name = function_name;
+    return functions.insert(function);
+}
+
+PProf__Line *add_line(datadog_arena **arena, uint64_t function_id, int64_t lineno) noexcept {
+    PProf__Line tmpline = PPROF__LINE__INIT;
+    tmpline.function_id = function_id;
+    tmpline.line = lineno;
+
+    auto line = (PProf__Line *) datadog_arena_alloc(arena, sizeof(PProf__Line));
+    new(line) PProf__Line(tmpline);
+    return line;
+}
+
+}
+
+void pprof_exporter::operator()(const recorder::event_table_t &event_table, string_table &strings,
+                                system_clock::time_point start_time, system_clock::time_point stop_time) {
+    if (event_table.empty()) {
+        return;
+    }
+
+    datadog_arena *arena = datadog_arena_create(4096);
+    pprof_set<PProf__Function> functions{&arena};
+    pprof_set<PProf__Location> locations{&arena};
+
+    // wall-time
+    // "thread id" "thread native id" as labels
+
+    std::stringstream ss{};
+    ss << "profile-" << num++ << ".pb";
+
+    for (auto &event : event_table) {
+        PProf__Sample sample = PPROF__SAMPLE__INIT;
+        sample.n_location_id = event->frames.size();
+        sample.location_id =
+            (uint64_t *) datadog_arena_alloc(&arena, sizeof(uint64_t) * sample.n_location_id);
+        unsigned local_locations = 0;
+
+        for (auto &frame : event->frames) {
+            uint64_t function_id = add_function(&arena, functions, frame.function_name);
+            PProf__Line *line = add_line(&arena, function_id, frame.lineno);
+
+            PProf__Location location = PPROF__LOCATION__INIT;
+            location.line = &line;
+            location.n_line = 1;
+            uint64_t location_offset = locations.insert(location);
+
+            sample.location_id[local_locations++] = locations[location_offset].id;
+            sample.value = (int64_t *) datadog_arena_alloc(&arena, sizeof(int64_t));
+            *sample.value = 0; // todo: this is nanoseconds since... what?
```

Comment:
If this is the wall time "data set", what we do in the python profiler is allocate the whole slice of time since the last sample to this one- https://github.com/DataDog/dd-trace-py/blob/master/ddtrace/profiling/collector/stack.pyx#L484, https://github.com/DataDog/dd-trace-py/blob/master/ddtrace/profiling/collector/stack.pyx#L488, https://github.com/DataDog/dd-trace-py/blob/master/ddtrace/profiling/collector/stack.pyx#L339.